atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java

```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+        assertThat(tokenRanges).hasSize(partitions);
+        assertThat(cassandra.refreshTokenRangesAndGetServers(() -> tokenRanges))
+                .hasSize(nodes.size())
+                .satisfies(servers -> {
+                    assertThat(Collections2.transform(servers, CassandraServer::cassandraHostName))
+                            .containsExactlyInAnyOrderElementsOf(
+                                    Collections2.transform(nodes, CassandraServer::cassandraHostName));
+                    assertThat(Collections2.transform(servers, CassandraServer::reachableProxyIps))
+                            .allSatisfy(ips -> assertThat(ips).allSatisfy(address -> {
+                                assertThat(address).isNotNull();
+                                assertThat(address.isUnresolved()).isFalse();
+                            }));
+                });
+    }
+
+    @Test
+    public void testComplexMultiDatacenterMultiRackPartialRefreshTokenRanges() {
+        int replicas = 3;
+        int partitions = 5;
+        Set<CassandraServer> nodes = generateNodes(replicas, partitions);
+        assertThat(nodes).allSatisfy(node -> assertThat(node.reachableProxyIps())
+                .allSatisfy(address -> assertThat(address.isUnresolved()).isTrue()));
+        CassandraService cassandra = clientPoolWithServers(nodes);
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+        assertThat(tokenRanges).hasSize(partitions);
+        assertThat(cassandra.refreshTokenRangesAndGetServers(() -> tokenRanges.subList(0, 3)))
+                .hasSize(9)
+                .satisfies(servers -> {
+                    assertThat(Collections2.transform(servers, CassandraServer::cassandraHostName))
+                            .containsAnyElementsOf(Collections2.transform(nodes, CassandraServer::cassandraHostName));
+                    assertThat(Collections2.transform(servers, CassandraServer::reachableProxyIps))
+                            .allSatisfy(ips -> assertThat(ips).allSatisfy(address -> {
+                                assertThat(address).isNotNull();
+                                assertThat(address.isUnresolved()).isFalse();
+                            }));
+                });
+    }
+
+    private static Set<CassandraServer> generateNodes(int replicas, int partitions) {
+        int nodeCount = replicas * partitions + 1;
+        assertThat(nodeCount).isLessThan(255);
+        return IntStream.range(1, nodeCount)
+                .mapToObj(i -> "10." + i + ".0.0")
+                .map(ip -> InetSocketAddress.createUnresolved(ip, DEFAULT_PORT))
+                .map(CassandraServer::of)
+                .collect(Collectors.toSet());
+    }
+
+    private static List<TokenRange> generateTokenRanges(int replicas, int partitions, Set<CassandraServer> nodes) {
```

Comment:
hmm this is a lot of code and I am afraid this needs tests of its own. I think it might be worth having token ranges hard coded in files and reading from them